### ISSUE TITLE
Upcoming breaking changes upstream in Org

### ISSUE THREAD

**Issue by yantar92** [ndwarshuis/om.el#46]
Please take a look at https://list.orgmode.org/874jnudps5.fsf@localhost/T/#u
The upcoming changes to Org element API are likely going to break org-ml as the internal representation of Org parse tree is going to be altered.

**Comment by ndwarshuis** [comment]
@yantar92 thanks for pointing this out. This will probably be a pain in the butt to update but it looks like org-ml itself should still function in more-or-less the same way. Overall it should also solve one of the biggest drawbacks to using `org-ml` (it's slow, especially without laziness).

**Comment by yantar92** [comment]
Nate Dwarshuis ***@***.***> writes:

> @yantar92 thanks for pointing this out. This will probably be a pain in the butt to update but it looks like org-ml itself should still function in more-or-less the same way. Overall it should also solve one of the biggest drawbacks to using `org-ml` (it's slow, especially without laziness).

I tried hard to reduce the amount of breakage, and even implemented some
built-in alternatives to org-ml APIs. In any case, feel free to join the
discussion at the mailing list if you think that some design
decisions are not good. I expect org-ml to be a large user of the new
API, apart from Org itself.

AFAIU, it may be enough on your side to update everything that is
setting and traversing property plists explicitly. `org-element-create'
and `org-element-properties-map' should hopefully cover all the needs.

-- 
Ihor Radchenko // yantar92,
Org mode contributor,
Learn more about Org mode at <https://orgmode.org/>.
Support Org development at <https://liberapay.com/org-mode>,
or support my work at <https://liberapay.com/yantar92>

**Comment by yantar92** [comment]
The changes have been merged onto Org's main branch.

**Comment by ndwarshuis** [comment]
thanks, I haven't had time to implement the changes yet but I've run all my tests against the new branch and everything "works" (which I think means it's mimicking whatever the new data structures are supposed to be without actually using them).

**Comment by yantar92** [comment]
Nate Dwarshuis ***@***.***> writes:

> thanks, I haven't had time to implement the changes yet but I've run all my tests against the new branch and everything "works" (which I think means it's mimicking whatever the new data structures are supposed to be without actually using them).

It might be an illusion.
The biggest change is that in (:standard-properties [...] :begin 1 ...),
the value of :standard-properties takes precedence.

However, (:begin 1 ...) will still work.

So, if you have places where you rely upon plist-get to retrieve or
modify the value, things may be broken when you are re-using the output
of `org-element-parse-buffer'/`org-element-at-point'.

Similarly, parser output may sometimes not produce all the properties
immediately, but calculate them on demand, later.

So, watch out for subtle bugs.

-- 
Ihor Radchenko // yantar92,
Org mode contributor,
Learn more about Org mode at <https://orgmode.org/>.
Support Org development at <https://liberapay.com/org-mode>,
or support my work at <https://liberapay.com/yantar92>

**Comment by ndw** [comment]
Am I getting bitten by this, or by something else? Background: I recently rebuilt my emacs config, letting `straight.el` pull the latest of most things, including `Org`. The breakage I'm seeing is that in a simple org file:

```
#+title This is a title
#+author Norm

* This is a heading
:PROPERTIES:
:CUSTOM_ID: h-E665E867-1813-4525-8711-23E04E6C3EEC
:END:

** TODO This is a todo
DEADLINE: <2023-08-22 Tue>
:PROPERTIES:
:CUSTOM_ID: h-A1FD3309-0340-4F9A-808B-6A5764F26766
:END:

This isn’t.

I’m going to attempt to get used to visual line mode because apparently that’s how modern editing tools work. Even though it bugs the living hell out of me because why are all the lines so long?

This is “test”.

#+BEGIN_SRC elisp
;; doesn’t really matter
Test "this". And 'this' and ... that. "More 'testing'". ..."test
#+END_SRC

** Second level
:PROPERTIES:
:CUSTOM_ID: h-C67D28D9-0878-4EC3-912F-38C58CE4E734
:END:

What about this?
```

If I run `(org-ml-parse-element-at (point-min))`, I get:

```
Debugger entered--Lisp error: (wrong-type-argument integer-or-marker-p nil)
  org-element--parse-elements(nil nil first-section nil nil nil (org-data nil))
  org-ml--parse-elements(nil nil first-section)
  org-ml--parse-element-at(1 nil)
  org-ml-parse-element-at(1)
  eval-expression((org-ml-parse-element-at (point-min)) nil nil 127)
  funcall-interactively(eval-expression (org-ml-parse-element-at (point-min)) nil nil 127)
  command-execute(eval-expression)
```

Which is a bit frustrating :-)

**Comment by ndwarshuis** [comment]
@ndw which specific version of org and emacs are you running? This still works for me on 9.6.6 (which doesn't appear to have this new api yet)

**Comment by ndw** [comment]
Straight seems to be giving me `9.7-pre` which is a bit of a surprise. Apologies for not noticing that. It's usually been safe to get the latest org, but perhaps that's not the case right now.

**Comment by yantar92** [comment]
Norman Walsh ***@***.***> writes:

> Straight seems to be giving me `9.7-pre` which is a bit of a surprise. Apologies for not noticing that. It's usually been safe to get the latest org, but perhaps that's not the case right now.

9.7-pre is development version. Sometimes, we make breaking changes,
like this one.

-- 
Ihor Radchenko // yantar92,
Org mode contributor,
Learn more about Org mode at <https://orgmode.org/>.
Support Org development at <https://liberapay.com/org-mode>,
or support my work at <https://liberapay.com/yantar92>

**Comment by ndw** [comment]
That's fair. I thought (I have a vague recollection) that dev work was in a branch so that what straight would get is the latest stable release. But maybe that's not the case. Simply backing out to 9.6.6, bundled with 29.1, is raising all kinds of other problems (to do with an apparent mixture of org versions) that I've never seen before. Sigh. But PEBKAC.

**Comment by ndw** [comment]
I haven't been this stymied by an Emacs configuration problem in a *long* time. Despite the fact that I've already used `(require 'org)` to make sure Org was loaded, when straight attempts to get `org-ml` it decides it needs to clone `org` as well and then it all falls apart.

**Comment by yantar92** [comment]
Norman Walsh ***@***.***> writes:

> That's fair. I thought (I have a vague recollection) that dev work was
> in a branch so that what straight would get is the latest stable
> release. But maybe that's not the case.

It is not. Straight is using the cutting edge - main branch.
You need "bugfix" to get stable-ish (it will still be stable pre-release).

> ... Simply backing out to 9.6.6, bundled with 29.1, is raising all kinds of other problems (to do with an apparent mixture of org versions) that I've never seen before. Sigh. But PEBKAC.

Yup. Because we revealed a bug in straight with recent changes:
https://github.com/radian-software/straight.el/issues/1107

-- 
Ihor Radchenko // yantar92,
Org mode contributor,
Learn more about Org mode at <https://orgmode.org/>.
Support Org development at <https://liberapay.com/org-mode>,
or support my work at <https://liberapay.com/yantar92>

**Comment by ndw** [comment]
For anyone else who stumbles down this path, what finally worked for me was adding
```
(add-to-list 'straight-built-in-pseudo-packages 'org)
(require 'org)
```
before straight started trying to fuss about with Org packages.

**Comment by neeasade** [comment]
`org-ml-update-headlines` seems broken for me in org 9.7, it wipes away the `* title` part of headlines (downgrading to builtin 9.6.x with the straight.el note above fixes for me)

**Comment by ndwarshuis** [comment]
@neeasade I'm currently working on fixing 9.7. It's expected that most things will break

**Comment by ndwarshuis** [comment]
I have a working (as in, passes all my tests) branch on update_org_9_7.

Breaking changes largely fall into several buckets:

1. changes that needed to be made due to things that were "wrong"
2. updates to handle new things in org 9.7, largely performance related but also other property nuances that are now in the element API (like timestamp deadlines)

I tried making aliases/wrapper functions in cases where I depreciated something, although its likely I missed a few.

Upside - the entire package is way faster thanks to the lazy eval and array properties in the new AST

Please test the new branch and let me know if anything unexpected breaks (see the changelog)

**Comment by neeasade** [comment]
Hi! Just want to report that I'm running the `update_org_9_7` branch successfully :saluting_face:

**Comment by neeasade** [comment]
minor nit: `org-ml-headline-get-planning` seems to be a little slow (I see there's a comment left [here](https://github.com/ndwarshuis/org-ml/blob/7154cb381222646e673b34f85161c13c19df3d15/org-ml.el#L5616-L5622) about it)

**Comment by ndwarshuis** [comment]
@neeasade yeah, this is a consequence to the new way everything under the headline is handled, which isn't 9.7-specific but I happened to notice the problem as I was updating

The basic, seemingly small but consequential problem is that the :pre-blank (the first space after the headline and everything else) is part of the headline itself and not included in anything below it. However, if a planning or drawer is added, then the :pre-blank needs to be added to :post-blank of the planning/drawer and removed from the headline.

This logic ended up being so complicated that I made a common interface for everything (called "supercontents") and based all the headline-get/set/map-X functions on this interface. The upshot is this reduces tons of code duplication, at the (noted) cost of speed, since this interface needs to generalize.

To get speed back, the options are:

1. duplicate code in the case of some of these simple getter functions
2. make some sort of lazy eval mechanism, which would work at the cost of lots of complexity

I'm not mortally opposed to doing (1) for the planning or node-property getters since those are easy to access and don't require me messing around with :pre-blank. Anything else I don't think is worth it.

(2) I'll probably eventually add, if only because it sounds fun :)

**Comment by ndwarshuis** [comment]
merged with master
